docs/dev/design/file-tree-diff.rst

```diff
@@ -0,0 +1,234 @@
+they work on the source files, and not on the generated files.
+
+All hope was lost for having nice features like this, until now.
+
+Proposed solution
+-----------------
+
+Since redirects and files of interest are related to the generated files.
+Instead of working over the source files, we will work over the generated files, which we have access to.
+
+The key points of this feature are:
+
+- Get the diff of the file tree between two versions.
+- Expose that as an API.
+- Integrate that in PR previews.
+
+Diff between two S3 directories
+-------------------------------
+
+We are already using ``rclone`` to speed up uploads to S3,
+``rclone`` has a command (``rclone check``) to return the diff between two directories.
+For this, it uses the metadata of the files, like size and hash
+(it doesn't download the files).
+
+.. code:: console
+
+   $ ls a
+   changed.txt  new.txt  unchanged.txt
+   $ ls b
+   changed.txt  deleted.txt  unchanged.txt
+   $ rclone check --combined=- /usr/src/app/checkouts/readthedocs.org/a /usr/src/app/checkouts/readthedocs.org/b
+   + new.txt
+   - deleted.txt
+   = unchanged.txt
+   * changed.txt
+
+The result is a list of files with a mark indicating if they were added, removed, or modified.
+The result is easy to parse.
+
+To start, we will only consider HTML files (``--include=*.html``).
+
+Lines changed between two files
+-------------------------------
+
+Once we have the list of files that changed, we can use a tool like ``diff`` to get the lines that changed.
+This is useful to link to the most relevant files that changed in a PR.
+
+.. code:: console
+
+   $ cat a.txt
+   One
+   Two
+   Three
+   Four
+   Five
+   $ cat b.txt
+   Ore
+   Three
+   Four
+   Five
+   Six
+   $ diff --side-by-side --suppress-common-lines a.txt b.txt
+   One                                                           | Ore
+   Two                                                           <
+                                                                 > Six
+
+.. note::
+
+   Taken from https://stackoverflow.com/questions/27236891/diff-command-to-get-number-of-different-lines-only.
+
+The command will return only the lines that changed between the two files.
+We can just count the lines, or maybe even parse each symbol to check if the line was added or removed.
+
+Another alternative is to use the `difflib <https://docs.python.org/3/library/difflib.html>`__ module,
+the only downside is that it doesn't distinguish lines that were changed from lines that were added or removed.
+But maybe that's ok? Do we really need to know if a line was changed instead of added or removed?
+
+.. code:: python
+
+   import difflib
+
+   diff = difflib.ndiff(["one", "two", "three", "four"], ["ore", "three", "four", "five"])
+   print(list(diff))
+   # ['+ ore', '- one', '- two', '  three', '  four', '+ five']
+
+A good thing of using Python is that we don't need to write the files to disk,
+and the result is easier to parse.
```

Comment:
It's probably slower as well, but I think I'm liking the idea of doing the diff in Python.